docs/source/notes/cpp-extensions.rst

```diff
@@ -0,0 +1,1020 @@
+C++ Extensions
+==============
+
+PyTorch provides a plethora of operations related to neural networks, arbitrary
+tensor algebra, data wrangling and other purposes. However, you may still find
+yourself in need of a more customized operation. For example, you might want to
+use a novel activation function you found in a paper, or implement an operation
+you developed as part of your research.
+
+The easiest way of integrating such a custom operation in PyTorch is to write it
+in Python by extending :class:`Function` and :class:`Module` as outlined `here
+<http://pytorch.org/docs/master/notes/extending.html>`_. This gives you the full
+power of automatic differentiation (spares you from writing derivative
+functions) as well as the usual expressiveness of Python. However, there may be
+times when your operation is better implemented in C++. For example, your code
+may need to be *really* fast because it is called very frequently in your model
+or is very expensive even for few calls. Another plausible reason is that it
+depends on or interacts with other C or C++ libraries. To address such cases,
+PyTorch provides a very easy way of writing custom *C++ extensions*.
+
+C++ extensions are a mechanism we have developed to allow users (you) to create
+PyTorch operators defined *out-of-source*, i.e. separate from the PyTorch
+backend. This approach is *different* from the way native PyTorch operations are
```

Comment:
Nit: I'm not sure how important it is for the reader to know how native PyTorch ops are implemented, except maybe as an aside.